packages/trackerless-network/src/logic/plumtree/PlumTreeManager.ts

```diff
@@ -0,0 +1,157 @@
+    }
+
+    onNeighborRemoved(nodeId: DhtAddress): void {
+        this.localPausedNeighbors.deleteAll(nodeId)
+        this.remotePausedNeighbors.deleteAll(nodeId)
+    }
+
+    async pauseNeighbor(node: PeerDescriptor, msgChainId: string): Promise<void> {
+        if (this.neighbors.has(toNodeId(node))) {
+            logger.debug(`Pausing neighbor ${toNodeId(node)}`)
+            this.remotePausedNeighbors.add(toNodeId(node), msgChainId)
+            const remote = this.createRemote(node)
+            await remote.pauseNeighbor(msgChainId)
+        }
+    }
+
+    async resumeNeighbor(node: PeerDescriptor, msgChainId: string, fromTimestamp: number): Promise<void> {
+        if (this.remotePausedNeighbors.isPaused(toNodeId(node), msgChainId)) {
+            logger.debug(`Resuming neighbor ${toNodeId(node)}`)
+            this.remotePausedNeighbors.delete(toNodeId(node), msgChainId)
+            const remote = this.createRemote(node)
+            await remote.resumeNeighbor(fromTimestamp, msgChainId)
+        }
+    }
+
+    getLatestMessageTimestamp(msgChainId: string): number {
+        if (!this.latestMessages.has(msgChainId) || this.latestMessages.get(msgChainId)!.length === 0) {
+            return 0
+        }
+        return this.latestMessages.get(msgChainId)![this.latestMessages.get(msgChainId)!.length - 1].messageId!.timestamp
+    }
+
+    async sendBuffer(fromTimestamp: number, msgChainId: string, neighbor: PeerDescriptor): Promise<void> {
+        const remote = new ContentDeliveryRpcRemote(this.localPeerDescriptor, neighbor, this.rpcCommunicator, ContentDeliveryRpcClient)
+        const messages = this.latestMessages.get(msgChainId)?.filter((msg) => msg.messageId!.timestamp > fromTimestamp) ?? []
+        await Promise.all(messages.map((msg) => remote.sendStreamMessage(msg)))
+    }
+
+    async onMetadata(msg: MessageID, previousNode: PeerDescriptor): Promise<void> {
+        // If we receive newer metadata than messages in the buffer, resume the sending neighbor
+        const latestMessageTimestamp = this.getLatestMessageTimestamp(msg.messageChainId)
+        if (latestMessageTimestamp < msg.timestamp) {
+            if (!this.metadataTimestampsAheadOfRealData.has(msg.messageChainId)) {
+                this.metadataTimestampsAheadOfRealData.set(msg.messageChainId, new Set())
+            }
+            this.metadataTimestampsAheadOfRealData.get(msg.messageChainId)!.add(msg.timestamp)
+            if (this.metadataTimestampsAheadOfRealData.get(msg.messageChainId)!.size > 1) {
+                console.log("resuming neighbor", latestMessageTimestamp, msg.timestamp, this.remotePausedNeighbors, this.localPausedNeighbors, this.neighbors.getAll().map((n) => toNodeId(n.getPeerDescriptor())))
+                await this.resumeNeighbor(previousNode, msg.messageChainId, this.getLatestMessageTimestamp(msg.messageChainId))
+                this.metadataTimestampsAheadOfRealData.get(msg.messageChainId)!.forEach((timestamp) => {
```

Comment:
I think there is a possible race condition here. More metadatas might arrive while the resume is still in progress and we would call the resume again (maybe on the same neighbor, maybe on another). Should we add a resumeInProgress flag?